update cyw20829 integration

npal-cy opened the pull request:
[Boards: cyw20829: add custom flashloaded with 64k of erase block size](https://github.com/zephyrproject-rtos/zephyr/commit/0483262cc1e3bae19c00298c39f9bfbc367747f7) 
- Add custom flashloaded (for openocd) with 64k of erase block size. Default one has 512, which cause MCUbootloader overwrite during flash primary application.

[soc: cyw20829: add support of Secure LCS](https://github.com/zephyrproject-rtos/zephyr/commit/b21a74522bfaf1ce99bca3effbe7a6f3091109b8) 

- Enable support of SECURE LCS stage. In this stage, the protection
 state is set to “secure”. A secured device will boot only when the authentication of its flash boot and application code succeeds.

[soc: cyw20829: Initial integrate Cypress MCUBoot for 20829](https://github.com/zephyrproject-rtos/zephyr/commit/b305e9fbda966bbc260aff6170ece734d303237c) 

- Added custom mcuboot cmake for sign/encrypt by using cysecuretools.

nordicjm commented on soc/infineon/cat1b/cyw20829/mcuboot.cmake:
```diff
@@ -0,0 +1,217 @@
+
+# This file includes extra build system logic that is enabled when
+# CONFIG_BOOTLOADER_MCUBOOT=y.
+#
+# It builds signed binaries using cysecuretools as a post-processing step
+# after zephyr/zephyr.elf is created in the build directory.
```
can you explain why this is needed?

npal-cy replied:
HI @nordicjm,
MCUboot support for CYW20829 device, has couple specifics:
- It supported only by Cypress branch of MCUboot https://github.com/mcu-tools/mcuboot/tree/cypress
>> we will provide documentation in board documentation how to build Cypress MCUboot bootloader (not Zephyr app) and operate with  with Zephyr (CONFIG_BOOTLOADER_MCUBOOT=y)

- In CYW20829 image,  encryption has some customization so we are using cysecuretools instead imagetool to sign/encrypt image, so [soc/infineon/cat1b/cyw20829/mcuboot.cmake](https://github.com/zephyrproject-rtos/zephyr/pull/81121/files#diff-601b7d4376d291f2ab0439b8a12eddb2be585adf8f023378ea7ede6f97cc292e) implements sign/encrypt by cysecuretools.

npal-cy commented:
Hi @carlescufi,
```
Binary file not allowed: boards/infineon/cyw920829m2evk_02/support/CYW208xx_SMIF_64K.FLM
Error: See https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style for more details.
```
we added openocd flashloader (boards/infineon/cyw920829m2evk_02) and  Compliance Checks fail as binary file not allowed.
Does Binary file rules also applies for openocd flashloader? 

Regards,
Nazar

npal-cy commented:
@talih0, @nordicjm, @ifyall can you review this PR?

nordicjm commented on soc/infineon/cat1b/cyw20829/mcuboot.cmake:
```diff
@@ -0,0 +1,217 @@
+# Copyright (c) 2024 Cypress Semiconductor Corporation.
```
it's a bit odd to see this copyright without the copyright from the `cmake/mcuboot.cmake` file given that I wrote a good portion of it and I can see this is a modified copy of that file

npal-cy commented on boards/infineon/cyw920829m2evk_02/board.cmake:
```diff
@@ -2,6 +2,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 board_runner_args(openocd "--target-handle=TARGET.cm33")
+board_runner_args(openocd "--openocd-search=${ZEPHYR_HAL_INFINEON_MODULE_DIR}/zephyr/blobs/flashloader/TARGET_CYW920829M2EVK-02")
```
Hi @nashif could you comment here?

Hi @nordicjm , could you point me to the documentation where described minimum requirements list?

Regards,
Nazar

nordicjm replied:
https://docs.zephyrproject.org/latest/contribute/bin_blobs.html#limited-scope

npal-cy replied:
hi @nordicjm 

i did updates. now we use custom flashloader (from blobs fetch) only for MCUboot case (where it is critical), for non MCUboot cases we use default one.

ifyall commented:
Issue [82922 ](https://github.com/zephyrproject-rtos/zephyr/issues/82922)was voted and approved 08 Jan 2025. This PR can be merged.

kartben commented:
@npal-cy needs a rebase please

npal-cy commented:
rebased, updated hal_infineon.

npal-cy commented:
I do not know why "Do Not Merge / Prevent Merging" still here ...

npal-cy commented:
hal_infineon is update to 59278e3598589a21d849e30baf3736954b4bc9ef, so removed DNM manually

nordicjm commented on dts/arm/infineon/cat1b/cyw20829/cyw20829.dtsi:
```diff
@@ -304,21 +339,21 @@
-					     <32 4>, /* CH13 */
-					     <33 4>, /* CH14 */
-					     <34 4>; /* CH15 */
+						 <20 4>, /* CH1 */
```
change seems unnecessary and wrong?

sreeramIfx replied:
updated this

nordicjm commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -18,3 +17,135 @@ zephyr_compile_definitions(CY_PDL_FLASH_BOOT)
+	# Cysecuretools policy.
+	if(NOT CONFIG_INFINEON_SECURE_POLICY)
+		# Get default cysecuretools policy
+
```
```suggestion
```

nordicjm replied:
not fixed

nordicjm commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -18,3 +17,135 @@ zephyr_compile_definitions(CY_PDL_FLASH_BOOT)
+
+		set(default_policy)
+		if(CONFIG_INFINEON_SECURE_LCS)
+			set(default_policy_name  policy_secure.json)
+		else()
+			set(default_policy_name  policy_no_secure.json)
```
single space not double between the value

nordicjm replied:
also cmake indent is 2 spaces

sreeramIfx replied:
updated..

nordicjm replied:
not fixed

nordicjm commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -18,3 +17,135 @@ zephyr_compile_definitions(CY_PDL_FLASH_BOOT)
+
+		# Try to find policy file in project folder
+		if("${default_policy}" STREQUAL "")
+			if (EXISTS ${APPLICATION_SOURCE_DIR}/policy/${default_policy_name})
```
no space between if and bracket, apply throughout

sreeramIfx replied:
updated throughout the file

nordicjm replied:
please fix all issues before re-requesting my review

sreeramIfx replied:
![image](https://github.com/user-attachments/assets/ba7f2e05-d150-4229-b30c-401dcbfe89e8)
@nordicjm - I am not seeing these issues in the latest file. Please let me know if I have not done the squash properly(when removing the fixup commits) and you are still seeing the older files?

nordicjm replied:
https://github.com/zephyrproject-rtos/zephyr/pull/81121/commits/3ef6ae18c4d90c8e9d8a48faa05617b4f3984d6c#diff-d8b6678de59c695a4ce45459480b1fb69d7d8819819e23515ef215edb18b5326R48

nordicjm commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -18,3 +17,135 @@ zephyr_compile_definitions(CY_PDL_FLASH_BOOT)
+			endif()
+		endif()
+
+		# Try to find policy file in Zephyr folder
+		if("${default_policy}" STREQUAL "")
+			if (EXISTS ${ZEPHYR_BASE}/policy/${default_policy_name})
+				set(default_policy ${ZEPHYR_BASE}/policy/${default_policy_name})
+			endif()
+		endif()
+
+		# Using default policy from cysecuretools targets
+		if("${default_policy}" STREQUAL "")
+			set(default_policy "${Python3_SITELIB}/cysecuretools/targets/cyw20829/policy/${default_policy_name}")
+		endif()
+	else()
+		if(IS_ABSOLUTE ${CONFIG_INFINEON_SECURE_POLICY})
+			if (NOT EXISTS  ${CONFIG_INFINEON_SECURE_POLICY})
+
+				message(FATAL_ERROR "Can't find policy: ${CONFIG_INFINEON_SECURE_POLICY}. "
+									"Please check if path in CONFIG_INFINEON_SECURE_POLICY is correct."
```
no, see how other files check multiple locations for files without duplicating the whole code multiple times e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/modules/extensions.cmake#L2725

sreeramIfx replied:
Updated logic to use `find_file`

nordicjm commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -18,3 +17,135 @@ zephyr_compile_definitions(CY_PDL_FLASH_BOOT)
+	message("-- Using cysecuretools policy: ${cysecuretools_policy}")
+endif()
+
+# Copy Zephyr application hex
```
complete no to the below, zephyr creates the file and you are completely hacking it to overwrite it with new data. Create a new file and set `BYPRODUCT_KERNEL_SIGNED_HEX_NAME` instead to the *new* name file. Zephyr's signed (MCUboot) filename is `<kernel>.signed.hex` so should follow that

sreeramIfx replied:
Please let me know if I am understanding this correctly, Instead of modifying the zephyr generated zephyr.hex file, you are asking to create a separate file(say 20829_zephyr.hex) and create 20829_zephyr.signed.hex from 20829_zephyr.hex?

sreeramIfx replied:
@nordicjm Can you please clarify your comment here.

nordicjm replied:
there is a Kconfig for the kernel name, you can use that as a base and change it. See https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/mcuboot.cmake#L163

nordicjm commented on soc/infineon/cat1b/cyw20829/Kconfig:
```diff
@@ -15,3 +15,23 @@ config SOC_SERIES_CYW20829
+config INFINEON_SECURE_LCS
+	bool "SECURE LCS stage support"
+	help
+		Enable support of SECURE LCS stage. In this stage, the protection
```
help indent is 1x tab followed by 2x spaces

sreeramIfx replied:
updated

nordicjm commented on soc/infineon/cat1b/cyw20829/Kconfig:
```diff
@@ -15,3 +15,23 @@ config SOC_SERIES_CYW20829
 	select SOC_EARLY_INIT_HOOK
+
+config INFINEON_SECURE_LCS
+	bool "SECURE LCS stage support"
```
```suggestion
	bool "Secure LCS stage support"
```

nordicjm commented on soc/infineon/cat1b/cyw20829/Kconfig:
```diff
@@ -15,3 +15,23 @@ config SOC_SERIES_CYW20829
+		state is set to “secure”. A secured device will boot only when the
+		authentication of its flash boot and application code succeeds.
+
+config INFINEON_SECURE_POLICY
+	string "Path to policy JSON file"
+	help
+		Policy is a text file in JSON format that contains a set of properties
+		for the device configuration (e.g., enabling/disabling debug access ports,
+		SMIF configuration, keys information, etc).
```
why is this in kconfig? Use a cmake variable

sreeramIfx replied:
I see other paths being KCONFIG symbols, https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/Kconfig.mcuboot#36. Why cmake variable?

nordicjm commented on soc/infineon/cat1b/cyw20829/Kconfig.soc:
```diff
@@ -26,6 +26,26 @@ config SOC_PACKAGE_CYW20829_40_QFN
+	hex
+	default $(dt_nodelabel_reg_addr_hex,sram_cbus)
+
+config CYW20829_ROM_START_OFFSET
+	hex
+	default 0x0
```
what? Why is this needed? Use the existing symbol

sreeramIfx replied:
removed CYW20829_ROM_START_OFFSET

nordicjm commented on boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts:
```diff
@@ -89,45 +90,50 @@ uart2: &scb2 {
+	flash0: flash@8000000 {
+		compatible = "soc-nv-flash";
+		reg = <0x08000000 DT_SIZE_M(1)>;
+
```
```suggestion
```

nordicjm commented on boards/infineon/cyw920829m2evk_02/cyw920829m2evk_02.dts:
```diff
@@ -89,45 +90,50 @@ uart2: &scb2 {
-		reg = <0x40890000 0x30000>;
+
+&qspi_flash {
+
```
```suggestion
```

nordicjm commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -149,3 +149,9 @@ set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
 	)
+
+# Use custom mcuboot cmake for sign/encrypt by using cysecuretools
+if (CONFIG_BOOTLOADER_MCUBOOT)
```
as previous comment

sreeramIfx replied:
removed space after if throughout the file

nordicjm commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -149,3 +149,9 @@ set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+# Use custom mcuboot cmake for sign/encrypt by using cysecuretools
+if (CONFIG_BOOTLOADER_MCUBOOT)
+  set_target_properties(zephyr_property_target PROPERTIES SIGNING_SCRIPT
+						${CMAKE_CURRENT_LIST_DIR}/mcuboot.cmake)
```
align properly

sreeramIfx replied:
updated

nordicjm commented on soc/infineon/cat1b/cyw20829/mcuboot.cmake:
```diff
@@ -0,0 +1,218 @@
+      endif()
+    endif()
+  endforeach()
+
+  #message(INFO " -- FLASH_SAHB: ${CONFIG_CYW20829_FLASH_SAHB_ADDR} / ${flash_addr_sahb_offset}")
+  #message(INFO " -- FLASH_CBUS: ${CONFIG_CYW20829_FLASH_CBUS_ADDR} / ${flash_addr_sbus_offset}")
```
to go

sreeramIfx replied:
removed the commented code

nordicjm commented on soc/infineon/cat1b/cyw20829/mcuboot.cmake:
```diff
@@ -0,0 +1,218 @@
+  # Basic 'cysecuretools' command and output format independent arguments.
+  set(cysecuretools_cmd ${CYSECURETOOLS}
+      -q -t cyw20829 -p ${cysecuretools_policy}
+    )
```
wrong alignment of end bracket

sreeramIfx replied:
updated

nordicjm commented on soc/infineon/cat1b/cyw20829/mcuboot.cmake:
```diff
@@ -0,0 +1,218 @@
+    # Use UNIX_COMMAND syntax for uniform results across host
+    # platforms.
+    separate_arguments(cysecuretools_extra_args UNIX_COMMAND
+                        ${CONFIG_MCUBOOT_EXTRA_IMGTOOL_ARGS})
```
fix alignment of ALL second lines which changes throughout most lines in this PR

sreeramIfx replied:
updated

nordicjm commented on soc/infineon/cat1b/cyw20829/mcuboot.cmake:
```diff
@@ -0,0 +1,218 @@
+  endif()
+
+  set_property(GLOBAL APPEND PROPERTY extra_post_build_byproducts ${byproducts})
+
```
```suggestion
```

nordicjm commented on soc/infineon/cat1b/cyw20829/mcuboot.cmake:
```diff
@@ -0,0 +1,218 @@
+
+  # Use overwrite-only instead of swap upgrades.
+  if(CONFIG_MCUBOOT_IMGTOOL_OVERWRITE_ONLY)
+    set(sign_image_cmd_args  ${sign_image_cmd_args} --overwrite-only --align 1)
```
2 space between arguments has crept in here, fix throughout whole PR

sreeramIfx replied:
updated throughout the PR

nordicjm commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -136,6 +136,79 @@ The path to the installed Infineon OpenOCD executable must be available to the `
+Operate in SECURE Lifecycle Stage
+*********************************
+
+The device lifecycle stage (LCS) is a key aspect of the security of the AIROC™ CYW20829 Bluetooth® MCU. The lifecycle stages follow a strict, irreversible progression dictated by the programming of the eFuse bits (changing the value from "0" to "1"). This system is used to protect the device's data and code at the level required by the user.
```
100 character width for documentation (excluding doxygen parts)

nordicjm commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -136,6 +136,79 @@ The path to the installed Infineon OpenOCD executable must be available to the `
+
+The following configuration options can be used to build for a device which has been provisioned to SECURE LCS and configured to use an encrypted flash interface:
+
+- ``CONFIG_INFINEON_SECURE_LCS``: used if the target device is in SECURE LCS
+- ``CONFIG_INFINEON_SECURE_POLICY``: path to the policy JSON file, which was created for provisioning the device to SECURE LCS (refer to section 3.2 "Key creation" of `AN239590 Provision CYW20829 to SECURE LCS`_)
+- ``CONFIG_INFINEON_SMIF_ENCRYPTION``: used if the device was configured to use an encrypted flash interface when provisioned to SECURE LCS.
```
use proper doxygen statements

sreeramIfx replied:
updated

nordicjm reviewed (commented):
Fixups not allowed, fix the original commits so the PR can be reviewed

kartben commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -69,26 +80,38 @@ Here is an example for building the :zephyr:code-sample:`blinky` sample applicat
 Programming and Debugging
 *************************
 
-.. zephyr:board-supported-runners::
```
Don't drop this? I'm guessing this got a lost during during the resolution of a merge conflict :)

sreeramIfx replied:
thank you, will add this back.

sreeramIfx committed:
boards: cyw20829: add custom flashloader with 64k of erase block size

Add custom flashloader (for openocd) with 64k of erase block size.
Default one has 512, which cause MCUbootloader overwrite during flash
primary application.

Signed-off-by: Sreeram Tatapudi <sreeram.praveen@infineon.com>

sreeramIfx commented:
@nordicjm Please review

nordicjm commented:
> @nordicjm Please review

You should click the re-request review buttons next to the names in the review list, otherwise this comment could get easily missed like just now when I lost this PR and suddenly remembered then had to search through every PR I've reviewed looking for "cy" to find it

nordicjm commented on boards/infineon/cyw920829m2evk_02/board.cmake:
```diff
@@ -2,6 +2,19 @@
 # SPDX-License-Identifier: Apache-2.0
 
 board_runner_args(openocd "--target-handle=TARGET.cm33")
+
```
commit typo *flashloader

nordicjm commented on soc/infineon/cat1b/cyw20829/Kconfig.soc:
```diff
@@ -26,6 +26,22 @@ config SOC_PACKAGE_CYW20829_40_QFN
 config SOC_PACKAGE_CYW20829_77_BGA
 	bool
 
+config CYW20829_FLASH_SAHB_ADDR
+	hex
+	default $(dt_nodelabel_reg_addr_hex,flash_sahb)
+
+config CYW20829_FLASH_CBUS_ADDR
+	hex
+	default $(dt_nodelabel_reg_addr_hex,flash_cbus)
+
+config CYW20829_SRAM_SAHB_ADDR
+	hex
+	default $(dt_nodelabel_reg_addr_hex,sram_sahb)
+
+config CYW20829_SRAM_CBUS_ADDR
+	hex
+	default $(dt_nodelabel_reg_addr_hex,sram_cbus)
```
these need to move to `Kconfig` file as they have no use in sysbuild (and there is no devicetree even)

nordicjm commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -30,6 +30,78 @@ math(EXPR flash_addr_offset
+
+if(CONFIG_INFINEON_SECURE_LCS OR CONFIG_BOOTLOADER_MCUBOOT)
+  # Check cysecuretools
+  find_program(CYSECURETOOLS cysecuretools)
```
should just use the required flag in find_program() rather then re-inventing the wheel

nordicjm commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -30,6 +30,78 @@ math(EXPR flash_addr_offset
+    default_policy
+    NAMES
+        ${CONFIG_INFINEON_SECURE_POLICY}
+        ${default_policy_name}
```
2 space indent not 4 for cmake

nordicjm commented on soc/infineon/cat1b/cyw20829/CMakeLists.txt:
```diff
@@ -30,6 +30,78 @@ math(EXPR flash_addr_offset
+    )
+
+  if(NOT default_policy)
+      message(FATAL_ERROR "Can't find policy:${CONFIG_INFINEON_SECURE_POLICY}"
```
missing space after :

nordicjm commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -117,7 +142,101 @@ The path to the installed Infineon OpenOCD executable must be available to the `
+The following configuration options can be used to build for a device which has been provisioned
+to SECURE LCS and configured to use an encrypted flash interface:
+
+- ``CONFIG_INFINEON_SECURE_LCS=y``: Enable if the target device is in SECURE LCS
```
use
```
:kconfig:option:`CONFIG_blah...`
```
and just say to set them not with the =blah part, so that they have links to the actual Kconfig descriptions

nordicjm commented on boards/infineon/cyw920829m2evk_02/doc/index.rst:
```diff
@@ -117,7 +142,101 @@ The path to the installed Infineon OpenOCD executable must be available to the `
+   :west-args: -p always
+   :gen-args: -DCONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_MCUBOOT_SIGNATURE_KEY_FILE=\"/path/to/cypress/mcuboot/boot/cypress/keys/cypress-test-ec-p256.pem\"
+
+If you use ``CONFIG_MCUBOOT_ENCRYPTION_KEY_FILE`` to generate an encrypted image then the final
```
as above

sreeramIfx commented:
> > @nordicjm Please review
> 
> You should click the re-request review buttons next to the names in the review list, otherwise this comment could get easily missed like just now when I lost this PR and suddenly remembered then had to search through every PR I've reviewed looking for "cy" to find it

I am not seeing that button on this PR, Probably because I was not the initial author of this PR? 
![image](https://github.com/user-attachments/assets/48bf3777-6af1-45b7-8667-b318549b44a7)

sreeramIfx committed:
soc: cyw20829: Initial integrate Cypress MCUBoot for 20829

Added custom mcuboot cmake for sign/encrypt by using cysecuretools

Signed-off-by: Sreeram Tatapudi <sreeram.praveen@infineon.com>
Signed-off-by: Nazar Palamar <nazar.palamar@infineon.com>

sreeramIfx committed:
soc: cyw20829: Use python script to generate app header

Instead of using app_header.c generate the app header using python
script and merge with final binary post build

Signed-off-by: Sreeram Tatapudi <sreeram.praveen@infineon.com>

sreeramIfx committed:
soc: cyw20829: add support of Secure LCS

Enable support of SECURE LCS stage. In this stage, the protection
state is set to “secure”. A secured device will boot only when the
authentication of its flash boot and application code succeeds

Signed-off-by: Sreeram Tatapudi <sreeram.praveen@infineon.com>

kartben reviewed (approved):
doc changes look ok - thanks!
